### DISCUSSION TITLE
Yggdrasil worker should send some event messages, when the worker is starting and stopping.

### DISCUSSION THREAD

**Discussion by jirihnidek** [RedHatInsights/yggdrasil#169] [General]
This discussion is motivated by this issue https://github.com/RedHatInsights/yggdrasil/issues/135

When yggdrasil worker is starting then, there is no handshake or hello message from worker to `yggd`. It is confusing, when you are developer of yggdrasil worker and you follow instructions from https://github.com/RedHatInsights/yggdrasil/blob/main/CONTRIBUTING.md. When you start worker at terminal all you get is this:

```
[jhnidek@thinkpad-p1 yggdrasil]$ go run ./worker/echo -log-level trace
2023/10/31 14:58:32 connecting to session bus: unix:path=/run/user/1000/bus
```

When there is no communication between `yggd` and worker, then you have no confidence that your worker works as expected. 

I would like to see there some events during start of worker (similar events that we send during processing dispatched message). Something like this during start on terminal with worker:

```
2023/10/31 14:59:10 emitting event STARTING
2023/10/31 14:59:11 emitting event STARTED
```

Following messages on the terminal with `yggd`:

```
[yggd] 2023/10/31 14:59:10 /home/jhnidek/github/redhat-insights/yggdrasil/cmd/yggd/client.go:1234: emitted event: {Worker:echo Name:STARTING MessageID:fff7d594-dc94-4a9f-8628-c60f9606ae54 ResponseTo: Data:map[]}
[yggd] 2023/10/31 14:59:11 /home/jhnidek/github/redhat-insights/yggdrasil/cmd/yggd/client.go:12345: emitted event: {Worker:echo Name:STARTED MessageID:fff7d594-dc94-4a9f-8628-c60f9606ae54 ResponseTo: Data:map[]}
```

Possibly something like this for the stopping of worker:

```
2023/10/31 15:09:03 emitting event STOPPING
2023/10/31 15:09:04 emitting event STOPPED
```

If you see events send from your worker and received by `yggd`, then you have confidence that they are able to talk to each other

Probably this is not the best solution for solving https://github.com/RedHatInsights/yggdrasil/issues/135, but it is something that could be useful for every developer of yggdrasil worker.

**Comment by m-horky** [comment]
I like the idea. Getting the list/count of connected workers dynamically (as you mentioned in the last paragraph) seems easier and less error-prone (the less state management the better), but I would expect the workers to announce their starts and exists.

**Comment by jirihnidek** [comment]
I realized that it will be enough to have less events, because during starting and stopping process nothing interesting could happen. The worked could not get blocked. Thus it would be sufficient to have only `STARTED` and `STOPPED` events.

**Comment by jirihnidek** [comment]
Related PR: https://github.com/RedHatInsights/yggdrasil/pull/187

**Comment by subpop** [comment]
I like the implementation in #187. I think that addresses this topic quite nicely.
